display/SINK.tally/tally_callbacks.cpp

```diff
@@ -0,0 +1,205 @@
+                                            "Devices", "Subdevices"),
+                            max_name_size);
+
+            for (const auto& [level,traffic]: data->traffic) {
+                std::string s = join_iterator(data->traffic_backend_name[level]);
+                print_compact("Explicit memory traffic (" + s + ")", traffic,
+                            std::make_tuple("Hostnames", "Processes", "Threads"),
+                            max_name_size);
+            }
+        }else {
+            for (const auto& [level,host]: data->host) {
+                std::string s = join_iterator(data->host_backend_name[level]); 
+                print_extended(s, host,
+                            std::make_tuple("Hostname", "Process", "Thread"),
+                            max_name_size);
+            }
+            print_extended("Device profiling", data->device,
+                            std::make_tuple("Hostname", "Process", "Thread",
+                                            "Device pointer", "Subdevice pointer"),
+                            max_name_size);
+
+            for (const auto& [level,traffic]: data->traffic) {
+                std::string s = join_iterator(data->traffic_backend_name[level]);
+                print_extended("Explicit memory traffic (" + s + ")", traffic,
+                            std::make_tuple("Hostname", "Process", "Thread"),
+                            max_name_size);
+            }
+        }
+    } else {
+        
+        nlohmann::json j;
+        j["units"] = {{"time", "ns"}, {"size", "bytes"}};
+        
+        if (data->display_metadata)
+            j["metadata"] = data->metadata;
+        
+        if (data->display_compact) {
+            for (auto& [level,host]: data->host)
+                j["host"][level] = json_compact(host);
+
+            if (!data->device.empty())
+                j["device"] = json_compact(data->device);
+
+            for (auto& [level,traffic]: data->traffic)
+                j["traffic"][level] = json_compact(traffic);
+
+        } else {
+            for (auto& [level,host]: data->host)
+                j["host"][level] = json_extented(host, std::make_tuple("Hostname", "Process", "Thread"));
+                
+            if (!data->device.empty())
+                j["device"] = json_extented(data->device,std::make_tuple(
+                    "Hostname", "Process","Thread", "Device pointer","Subdevice pointer"));
+
+            for (auto& [level,traffic]: data->traffic)
+                    j["traffic"][level] = json_extented(traffic,std::make_tuple(
+                        "Hostname", "Process", "Thread"));
+        }
+        std::cout << j << std::endl;
+    }
+
+    auto stop = high_resolution_clock::now();
+
+    // Subtract stop and start timepoints and
+    // cast it to required unit. Predefined units
+    // are nanoseconds, microseconds, milliseconds,
+    // seconds, minutes, hours. Use duration_cast()
+    // function.
```

Comment:
I think we need to explain "why" we need to cast, and not how we are doing it

Reply:
Removed, since we can time the code using the time command, maybe in further iterations we will need to time some sections, then I will implement some timing...